scipy/stats/_multivariate.py

```diff
@@ -3850,3 +3852,448 @@ def rvs(self, dim, size=1, random_state=None):
+    a :math:`2 \times 2` matrix. The result, in fact, has shape ``(2, 2, 2)``,
+    where ``result[0]`` is equal to
+    ``multivariate_hypergeom.cov(m=[7, 9], n=8)`` and ``result[1]`` is equal
+    to ``multivariate_hypergeom.cov(m=[10, 15], n=12)``.
+
+    Alternatively, the object may be called (as a function) to fix the `m`
+    and `n` parameters, returning a "frozen" multivariate hypergeometric
+    random variable.
+
+    >>> rv = multivariate_hypergeom(m=[10, 20], n=12)
+    >>> rv.pmf(x=[8, 4])
+    0.0025207176631464523
+
+    See also
+    --------
+    scipy.stats.hypergeom : The hypergeometric distribution.
+    numpy.random.Generator.multivariate_hypergeometric : The multivariate hypergeometric distribution sampler in numpy.
+    scipy.stats.multinomial : The multinomial distribution.
+
+    References
+    ----------
+    .. [1] The Multivariate Hypergeometric Distribution,
+           http://www.randomservices.org/random/urn/MultiHypergeometric.html
+
+    .. [2] Thomas J. Sargent and John Stachurski, 2020,
+           Multivariate Hypergeometric Distribution
+           https://python.quantecon.org/_downloads/pdf/multi_hyper.pdf
+    """  # noqa: E501
+    def __init__(self, seed=None):
+        super(multivariate_hypergeom_gen, self).__init__(seed)
+        self.__doc__ = doccer.docformat(self.__doc__, mhg_docdict_params)
+
+    def __call__(self, m, n, seed=None):
+        """
+        Create a frozen multivariate_hypergeom distribution.
+
+        See `multivariate_hypergeom_frozen` for more information.
+        """
+        return multivariate_hypergeom_frozen(m, n, seed=seed)
+
+    def _process_parameters(self, m, n):
+        m = np.asarray(m, dtype=np.int_)
+        n = np.asarray(n, dtype=np.int_)
+        if m.ndim == 0:
+            raise ValueError("'m' must be an array!")
+        mcond = np.any(m < 0, axis=-1)
+
+        M = m.sum(axis=-1)
+
+        ncond = (n <= 0) | (n > M)
+        return M, m, n, mcond | ncond
+
+    def _process_quantiles(self, x, M, m, n):
+        xx = np.asarray(x, dtype=np.int_)
+        if xx.ndim == 0:
+            raise ValueError("x must be an array.")
+        if xx.size != 0 and not xx.shape[-1] == m.shape[-1]:
+            raise ValueError(f"Size of each quantile should be size of 'm': "
+                             f"received {xx.shape[-1]}, "
+                             f"but expected {m.shape[-1]}.")
+
+        xcond = np.any(xx != x, axis=-1)
+        xcond = xcond | np.any((xx < 0) | (xx > m), axis=-1)
+        xcond = xcond | (xx.sum(axis=-1) != n)
+        return xx, xcond
+
+    def _checkresult(self, result, cond, bad_value):
+        result = np.asarray(result)
+        if cond.ndim != 0:
+            result[cond] = bad_value
+        elif cond:
+            return bad_value
+        if result.ndim == 0:
+            return result[()]
+        return result
+
+    def _logpmf(self, x, M, m, n):
+        num = betaln(m+1, 1) - betaln(x+1, m-x+1)
```

Comment:
Yes. When two or more negative values are present in quantiles, it raises a runtime warning. I thought about it but am unable to see an easy fix. I think `_lazywhere` or `np.where` won't work as the warning is raised by `np.sum` when it tries to add two `nan` values (generated due to negative x values). Any ideas?

Reply:
Hmm, and I think I got it for something else. Without thinking too carefully, maybe 

- `num` starts as an array of `nans`
- you generate a mask that distinguishes between which outputs _should_ be `nan` and which have a valid answer
- replace the elements of `num` that have a valid answer (using the mask)

Reply:
Yeah. This helped. `-np.inf` were causing the warning. I masked with np.nan and returned the final expected result.

Reply:
I have added a comment explaining from where this equation comes from.

Reply:
Hmm I'm still getting a RuntimeWarning for the first thing I tried:
```
from scipy.stats import multivariate_hypergeom, multinomial, multivariate_normal
res = multivariate_hypergeom.pmf(x=[1, 11], m=[10, 1], n=12)
print(res)
```
```
C:\Users\matth\Desktop\scipy\scipy\stats\_multivariate.py:4086: RuntimeWarning: invalid value encountered in subtract
  return np.where(cond, np.NINF, num) - den
```
and there's a different RuntimeWarning for
```
res = multivariate_hypergeom.pmf(x=[1, 11], m=[10, -1], n=12)
```
```
C:\Users\matth\Desktop\scipy\scipy\stats\_multivariate.py:4080: RuntimeWarning: invalid value encountered in subtract
  num = betaln(m+1, 1) - betaln(x+1, m-x+1)
```

Reply:
The former can be fixed but I am not sure how to fix the latter as it is caused by `betaln` function

Reply:
Yeah, the trick would be to use masking to ensure that betaln never gets those invalid values.

Reply:
> the trick would be to use loñgical indexing to ensure that betaln never sees those invalid values

I am not familiar with the code of `betaln` and "_loñgical indexing_". Can you elaborate more, please?

Reply:
Fixed the former runtime warning in the latest commit.

Reply:
Haha that was a typo. I meant "logical indexing" aka "boolean indexing" - using a boolean array to index a matrix. You are doing that after calling `betaln`, but I suspect we want to do it before calling `betaln` to pick out only the rows that won't cause the error. If this doesn't make sense, I'll look into it more closely. Perhaps there is something I'm missing.

Reply:
Here's a bit more detail using a made-up example.

Suppose `m` is an N-d array. I'll let it be 3 x 4 x 5.
```
import numpy as np
m = np.random.rand(3, 4, 5)
```
We want to call a function (e.g. `betaln`) on only the rows of `m` for which some condition holds (e.g. we won't get a `nan` and runtime error). Set up an array to hold the results:
```
res = np.zeros_like(m)
```
`i` will be the boolean array that selects the rows of `m` for which we'll get a valid result. In this case, I'm going to set up `i` arbitrarily:
```
i = m.sum(axis = -1) > 2.5 
```
Then we can perform whatever operation we want only on the selected rows.
```
res[i] = np.log(m[i])
```
Your case is a little more complicated, but it's the same principle. `i` is like `mcond` or `mncond` (haven't thought too carefully yet about which, or whether it's different), and instead of `np.log` you're doing `beta`. Actually, you probably want `res` to be a 2D array instead of shaped like `m`, so you do the sum before storing the results.  But I think with this sort of approach you can eliminate all the RuntimeWarnings.

Reply:
Ah, I get it now. Thanks for such an elaborate reply. I will try to do this and add tests with R's multivariate hypergeometric distribution so we are done with the major tasks remaining now.

Reply:
All the `RuntimeWarning`s fixed now! Thanks to your suggestion